### ISSUE TITLE
TCP Packet loss: Telepresence modify SNAT?

### ISSUE THREAD

**Issue by govargo** [telepresenceio/telepresence#1065]
### What were you trying to do?

I have PHP container on GKE.
I am using Telepresence to debug and develop my application.

I want to use debugger(Xdebug in PHP) to debug and develop more effectively.
I ran following command. 
But Xdebug did not work normally.

```
telepresence --mount=/tmp/known --swap-deployment k8s-laravel-app --expose 1008
0 --docker-run -u 33:33 -it --rm --privileged -p 10080:10080 laravel-app:customize-for-telepresence
```

### What did you expect to happen?

I want to debug to my application in docker.
I can confirm that I can debug application with Xdebug in case I use only normal docker.

But Xdebug is not working with using Telepresence.
So, I can't debug my container application.
(The details are described later)

### Constitution & Enviroment

PC: Mac OS 10.14.5
Python: 3.7.3
Telepresence 0.100
Container(Deployment Target): PHP 7.2.8 using Laravel Framework. Nginx + PHP-fpm with Xdebug
Kubernetes: 1.13.6-gke.13

#### First, The application is working properly with following command.
Nginx can receive access from ingress endpoint or local PC Browse(Google Chrome).

```
telepresence --mount=/tmp/known --swap-deployment k8s-laravel-app --expose 1008
0 --docker-run -u 33:33 -it --rm --privileged -p 10080:10080 laravel-app:customize-for-telepresence
```

#### Second, I inject Xdebug configfile to PHP Container.
I rebuild my image(`laravel-app:customize-for-telepresence`).
I ran above telepresence command again.

docker-php-ext-xdebug.ini is here.
```
zend_extension=xdebug.so
xdebug.remote_enable = 1
xdebug.remote_autostart = 0
xdebug.remote_host = "172.24.43.157"
xdebug.remote_port = 9000
xdebug.idekey = PHPSTORM
xdebug.remote_log=/var/log/xdebug.log
```

`172.24.43.157` is my host ipaddress of Mac PC.

I set breakpoint on IDE(VSCode).
when I access my local ipaddress `http://localhost:10080/`, then xdebug log says `Connected to client`.

here is xdebug.log
```
# cat /var/log/xdebug.log
[69] Log opened at 2019-06-28 11:08:43
[69] I: Connecting to configured address/port: 172.24.43.157:9000.
[69] I: Connected to client. :-)
[69] -> <init xmlns="urn:debugger_protocol_v1" xmlns:xdebug="https://xdebug.org/dbgp/xdebug" fileuri="file:///project/public/index.php" language="PHP" xdebug:language_version="7.2.8" protocol_version="1.0" appid="69" idekey="PHPSTORM"><engine version="2.7.2"><![CDATA[Xdebug]]></engine><author><![CDATA[Derick Rethans]]></author><url><![CDATA[https://xdebug.org]]></url><copyright><![CDATA[Copyright (c) 2002-2019 by Derick Rethans]]></copyright></init>
```

VSCode launch.json
```
{
     "name": "Listen for XDebug",
     "type": "php",
     "request": "launch",
     "port": 9000,
     "pathMappings": {
          "/app":"/Users/govargo/git/laravel-app"
     }
}
```

I access to my application, however, application process does not stop at breakpoint.

#### Third, I deep dive to solve this problem.

I enter into laravel-app container(laravel-app:customize-for-telepresence).
```
docker exec -it -u root 237f2e832047 bash
```

And I attach network-debugg container to laravel-app container in order to investigate.
```
docker run -it --rm --name bcc --privileged --network host \
   -v /var/run/docker.sock:/host/var/run/docker.sock -v /dev:/host/dev  \
   -v /usr:/host/usr:ro -v /etc:/host/etc:ro --pid container:237f2e832047 \
   --net container:237f2e832047 --entrypoint bash bcc
```

Here is network in container(laravel-app)
```
laravel container# ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.2  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:ac:11:00:02  txqueuelen 0  (Ethernet)
        RX packets 23379  bytes 3464321 (3.3 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 25033  bytes 3509904 (3.3 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1  (Local Loopback)
        RX packets 74550  bytes 7210333 (6.8 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 74550  bytes 7210333 (6.8 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

I run following command.
In network-debugg container: `tcpdump -i any -s 0 -nnn port 9000`
In host PC: `tcpdump -i any -s 0 -nnn port 9000`
In laravel container: `nc 172.24.43.157 9000`

`172.24.43.157` is my host ipaddress. and `9000` port is used by VSCode in host.

The Result is here.
```
laravel container# nc 172.24.43.157 9000

network-debugg container# tcpdump -i any -s 0 -nnn port 9000
listening on any, link-type LINUX_SLL (Linux cooked), capture size 262144 bytes
11:32:24.022415 IP 172.24.43.157.9000 > 172.17.0.2.56460: Flags [S.], seq 1440960005, ack 2167296949, win 43690, options [mss 65495,sackOK,TS val 13132839 ecr 13132839,nop,wscale 7], length 0

host PC# tcpdump -i any -s 0 -nnn port 9000
(output is nothing)
```

I am questioning the results `172.24.43.157.9000 > 172.17.0.2.56460`.
Considering the flow of the network, `container(172.17.0.2) → host(172.24.43.157)` is correct.
However, the result is reversed.

Though I do not know the root cause, I can confirm the TCP packet is lost.
So, TCP Connection cannot be established between IDE(VSCode) and Xdebug.

And I experimentally ran following command.
```
laravel container# nc -u 172.24.43.157 9000
a

network-debugg container# tcpdump -i any -s 0 -nnn port 9000
11:39:51.289230 IP 172.17.0.2.34297 > 172.24.43.157.9000: UDP, length 2

host PC# tcpdump -i any -s 0 -nnn port 9000
20:39:46.327312 IP 172.24.43.157.49943 > 172.24.43.157.9000: UDP, length 2
```

When I try UDP Connection, The Connection is established correctly!
The result is not reversed.

I suspect that this is a problem with SNAT in case of TCP.
So, I would like to confirm if Telepresence twist TCP communication.

Especially since communication from host to container is nothing of problem,
**I would like to know if Telepresence twist communication from container to host**.

### Added(2019/07/01)

Only Docker(without Telepresence), TCP Connection is correct.

```
laravel container# nc -u 172.24.43.157 9000

host PC# tcpdump -i any -s 0 -nnn port 9000
11:04:03.283783 IP 172.24.43.157.59908 > 172.24.43.157.9000: Flags [S], seq 1368874955, win 65535, options [mss 16344,nop,wscale 6,nop,nop,TS val 981538043 ecr 0,sackOK,eol], length 0
11:04:03.283794 IP 172.24.43.157.59908 > 172.24.43.157.9000: Flags [S], seq 1368874955, win 65535, options [mss 16344,nop,wscale 6,nop,nop,TS val 981538043 ecr 0,sackOK,eol], length 0
11:04:03.283851 IP 172.24.43.157.9000 > 172.24.43.157.59908: Flags [S.], seq 681438641, ack 1368874956, win 65535, options [mss 16344,nop,wscale 6,nop,nop,TS val 981538043 ecr 981538043,sackOK,eol], length 0
11:04:03.283854 IP 172.24.43.157.9000 > 172.24.43.157.59908: Flags [S.], seq 681438641, ack 1368874956, win 65535, options [mss 16344,nop,wscale 6,nop,nop,TS val 981538043 ecr 981538043,sackOK,eol], length 0
``` 

Thank you for reading.
I hope you have some advice or comments.

**Comment by ark3** [comment]
Thank you for the detailed report. In short, the container method sends all traffic to the cluster. That includes traffic that would otherwise reach the host. I think you would benefit from a fix for #1022.

**Comment by govargo** [comment]
@ark3 Thank you for your comment.

> the container method sends all traffic to the cluster. That includes traffic that would otherwise reach the host

I understood this behavior. And I checked #1022.
I'm looking forward to a fix for #1022.
